# TD 1 | Introduction à Python pour l'analyse de données

---

Objectifs du TD :

* découvrir Python
* se familiariser avec le langage et le notebook Jupyter
* découvrir et maîtriser les bases des librairies de calcul numérique et d'analyse de données numpy et pandas

---

##  Installation de Python et Jupyter

### Windows

#### Installation Python

Vérifiez si vous avez déjà une version de python installée en ouvrant l'invite de commande et en exécutant l'une des deux commandes python suivantes :
> python --version   

> python3 --version

Si ce n'est pas le cas, vous pouvez procéder à l'installation comme suit :

1. Allez sur le site web : https://www.python.org/downloads/windows/
2. Cliquez sur "Latest Python 3 Release - Python 3.10.7".
3. Si votre ordinateur utilise une version 64 bits de Windows, téléchargez le Windows installer (64-bit). Sinon, téléchargez Windows installer (32-bit).
4. Après avoir téléchargé le Windows installer, vous devez l'exécuter (double-cliquez dessus) et suivre les instructions qui s'y trouvent.

NB : Au début de l'installation, assurez-vous de cocher la case "Add Python 3.10 to PATH" ou "Add Python to your environment variables" avant de cliquer sur "Install Now".


#### Installation Jupyter

Ouvrir l'invite de commande et exécuter la commande suivante en remplaçant Pseudo_de_votre_user par le nom de l'utilisateur de votre session :
> cd "C:\Users\Pseudo_de_votre_user\AppData\Local\Programs\Python\Python310\Scripts"

Ensuite, executer la commande suivante :
> pip install jupyterlab

Une fois l'installation terminée avec succès, vous pouvez lancer JupyterLab avec la commande suivante :
> jupyter-lab


### Linux & Mac OS

#### Installation Python

Normalement, Python est préinstallé sur la quasi-totalité des distributions. Vous n'avez pas besoin de le réinstaller, mais vous pouvez mettre à jour la version si vous le souhaitez.

Vérifiez la version Python installée en exécutant l'une des deux commandes python suivantes sur votre terminal :
> python --version   

> python3 --version

#### Installation Jupyter

Vérifier que la commande pip3 est installée. Sinon, l'installer (exemple pour Debian/Ubuntu/Linux Mint, à adapter selon votre distribution) :

> sudo apt install python3-pip

Ensuite, installer JupyterLab en executant la commande suivante:
> pip3 install -U jupyterlab

Enfin, placez-vous dans votre dossier de travail et lancez un notebook :

> jupyter-lab

## Syntaxe et Concepts de base

In [1]:
a = 33
b = 9
c = 2.0

In [2]:
a+b

42

In [3]:
print(a+b)

42


In [4]:
type(a)

int

In [5]:
type(c)

float

In [6]:
a = 'hello'
type(a)

str

In [7]:
# Affiche les entiers pairs de 1 à 10
for i in range(1, 11):
    if i % 2 == 0:
        print(i)

2
4
6
8
10


### List

In [8]:
l = [1,2,3,4,4]

#### len, max, min

In [10]:
print('longueur de la liste =', len(l))
print("Valeur max de l =", max(l))
print("Valeur min de l =", min(l))

longueur de la liste = 5
Valeur max de l = 4
Valeur min de l = 1


#### Ajout d'un élément d'une liste

In [11]:
l.append(6)
l

[1, 2, 3, 4, 4, 6]

In [12]:
l = l+[7]
l

[1, 2, 3, 4, 4, 6, 7]

In [13]:
l+=[7]
l

[1, 2, 3, 4, 4, 6, 7, 7]

#### Modifier un élément

In [14]:
l[0] = 29
l

[29, 2, 3, 4, 4, 6, 7, 7]

In [15]:
l.insert(1,9)
l

[29, 9, 2, 3, 4, 4, 6, 7, 7]

#### Suppression des éléments d'une liste

In [16]:
l.remove(7)
l

[29, 9, 2, 3, 4, 4, 6, 7]

In [17]:
l.pop(2)
l

[29, 9, 3, 4, 4, 6, 7]

#### Trier une liste

In [18]:
l.sort(reverse=True)
print(l)

[29, 9, 7, 6, 4, 4, 3]


In [19]:
l_sorted = sorted(l)
print('Sorted list:', l_sorted)

Sorted list: [3, 4, 4, 6, 7, 9, 29]


#### Liste non homogène

In [20]:
l = [1, 'un', 2]
l.append('deux')
l += [3, 'trois']
print(type(l))
l

<class 'list'>


[1, 'un', 2, 'deux', 3, 'trois']

In [21]:
a = l[3], l[-1]
print(type(a))
a

<class 'tuple'>


('deux', 'trois')

In [22]:
a[0]

'deux'

In [23]:
[2**p for p in range(10)] # "Compréhension de liste"

[1, 2, 4, 8, 16, 32, 64, 128, 256, 512]

In [24]:
# Liste des entiers pairs de 1 à 10
l = [i for i in range(1, 11) if i % 2 == 0]
print(l)

[2, 4, 6, 8, 10]


### Tuple

In [25]:
p = tuple(range(1,5))
print("p=", p)
print("p[0] = ", p[0])

p= (1, 2, 3, 4)
p[0] =  1


In [26]:
p = ([1,2],5)

print(p)

p[0][0] = 8

print(p)

([1, 2], 5)
([8, 2], 5)


In [27]:
# Les éléments du tuple sont fixes
p[0] = 6

TypeError: 'tuple' object does not support item assignment

In [28]:
p1 = (1,2)
p2 = (2,1)

print(p1 == p2)

False


### Set

In [29]:
s = set((1,2,'l'))
print(s)

s = {4,5,9,9,9,9}
print(s)

{1, 2, 'l'}
{9, 4, 5}


In [30]:
#Ajout d'un élément au set
s.add(6)
print(s)

{9, 4, 5, 6}


In [31]:
#Suppression d'un élément du set
s.discard(5)
print(s)

{9, 4, 6}


In [32]:
s.clear()
print(s)

set()


In [33]:
A = {1,2,3,4,5}
B = {2,4,6,7}
C = {1,2,3}

In [34]:
#Intersection
print("Intersection de A et B =", A&B)

#Union
print("Union entre A et B =", A|B)

#Exclusion
print("Exclusion de B de A =", A-B)

#Différence symétrique
print("Diff sym. A et B =", A^B)

Intersection de A et B = {2, 4}
Union entre A et B = {1, 2, 3, 4, 5, 6, 7}
Exclusion de B de A = {1, 3, 5}
Diff sym. A et B = {1, 3, 5, 6, 7}


In [35]:
print("B est inclus strict dans A") if A>B else print(False)
print("C est inclus dans A") if A>=C else print(False)

False
C est inclus dans A


### Dictionary

In [36]:
# 1ère façon de créer un dictionnaire
dic = {"key1": "value1", "answer": 42}
# 2ème façon de créer un dictionnaire
dic2 = dict(key1="value1", answer=42)
# 3ème façon de créer un dictionnaire
dic3 = dict([("key1","value1"), ("answer",42)])
print(dic)
print(dic2)
print(dic3)

dic["answer"]

{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}


42

In [37]:
dic['new'] = [1,2,3]

In [38]:
dic.keys()

dict_keys(['key1', 'answer', 'new'])

In [39]:
dic.items()

dict_items([('key1', 'value1'), ('answer', 42), ('new', [1, 2, 3])])

In [40]:
dic.values()

dict_values(['value1', 42, [1, 2, 3]])

In [41]:
a,b,c = dic.items()
print(a)

('key1', 'value1')


## Programmation Orientée Objet

In [42]:
"""
Exemple de classe en Python
"""
class Moteur:

    # Constructeur
    def __init__(self, esn, panne=False):
        self.esn = esn
        self.panne = panne

    # Méthodes
    def dire_bonjour(self):
        print('Bonjour, mon numéro de série est ' + self.esn)

    def fonctionne(self):
        return not self.panne


In [43]:
mot1 = Moteur('420912')
mot2 = Moteur(panne=True, esn='420913')

mot1.dire_bonjour()
print(mot1.fonctionne())

mot1.panne = True
print(mot1.fonctionne())

print('\n')
mot2.dire_bonjour()
print(mot2.fonctionne())

Bonjour, mon numéro de série est 420912
True
False


Bonjour, mon numéro de série est 420913
False


### Méthode récurcive vs Méthode itérative

In [44]:
# Méthode récursive
def fact1(n):
    if n == 0:
        return 1
    else:
        return n*fact1(n-1)

# Méthode itérative
def fact2(n):
    s=1
    for i in range(1,n+1):
        s*=i
    return s

print(fact1(5))
print(fact2(5))

120
120


In [45]:
"""
App.1 : Ecrire une fonction qui retourne le plus grand commun diviseur (pgcd) entre deux nombres a et b.
"""

'\nApp.1 : Ecrire une fonction qui retourne le plus grand commun diviseur (pgcd) entre deux nombres a et b.\n'

In [46]:
"""
App.1 : Correction
"""

#Méthode itérative
def pgcd1(a,b):
    while (a%b) != 0:
        s = a%b
        a = b
        b = s
    return b


#Méthode récurcive
def pgcd2(a,b):
    if a==b : return a
    if a<b : a,b = b,a
    return pgcd2(a-b,b)

a = 230
b = 25

print("pgcd1(a,b) = ", pgcd1(a,b))
print("pgcd2(a,b) = ", pgcd2(a,b))

pgcd1(a,b) =  5
pgcd2(a,b) =  5


## Libraries

> pip install numpy pandas

> pip3 install -U numpy pandas

### Math

In [47]:
import math

In [48]:
x = 4

print("sqrt(x) =", math.sqrt(x))
print("cos(x) =", math.cos(x))
print("sin(x) =", math.sin(x))
print("factorial(x) =", math.factorial(x))

sqrt(x) = 2.0
cos(x) = -0.6536436208636119
sin(x) = -0.7568024953079282
factorial(x) = 24


### Numpy

In [49]:
import numpy as np

In [50]:
# Création d'un array à partir d'une liste
v = np.array([1.0, 2.0, 3.0])
# Création d'un array de taille (n,m) initialisé à 0
z = np.zeros((3,4))
z

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [51]:
# Taille d'un array
print(z.shape)
print(v.shape)

(3, 4)
(3,)


In [52]:
# Opérations courantes
A = np.ones((3,3)) + np.eye(3)
print('A = ', A)
print('Av = ', np.dot(A,v))
print('3*A = ', 3*A)
print('A*v = ', A*v)
print('A + 1 = ', A+1)
print('A + v = ', A+v)
print('A^2 = ', np.square(A))

A =  [[2. 1. 1.]
 [1. 2. 1.]
 [1. 1. 2.]]
Av =  [7. 8. 9.]
3*A =  [[6. 3. 3.]
 [3. 6. 3.]
 [3. 3. 6.]]
A*v =  [[2. 2. 3.]
 [1. 4. 3.]
 [1. 2. 6.]]
A + 1 =  [[3. 2. 2.]
 [2. 3. 2.]
 [2. 2. 3.]]
A + v =  [[3. 3. 4.]
 [2. 4. 4.]
 [2. 3. 5.]]
A^2 =  [[4. 1. 1.]
 [1. 4. 1.]
 [1. 1. 4.]]


In [53]:
B = np.array([5, 2, 9, 1])

print(np.sort(B))

[1 2 5 9]


#### Compatif de performance : produit matriciel


In [54]:
import random

In [55]:
# Création d'une matrice aléatoire sous forme de liste de listes
taille = 2
A = [[random.random() for _ in range(taille)] for _ in range(taille)]

In [56]:
print(A)

[[0.32595025458666016, 0.40314822207281487], [0.384854648781902, 0.34270022547229195]]


In [57]:
"""
EXERCICE - Afficher un tuple contenant les dimensions de la matrice A
"""

a = (len(A), len(A[0]))
a
# a = np.array(A).shape

(2, 2)

$$ (AB)_{ij} = \sum_k A_{ik} B_{kj} $$

In [59]:
"""
EXERCICE - Implémenter le produit matriciel de 2 matrices sous forme de listes de listes python
"""

def produit(A, B):
  rowsA = len(A); colsA = len(A[0])
  rowsB = len(B); colsB = len(B[0])

  prod = [[None for i in range(colsB)] for _ in range(rowsA)]

  for i in range(rowsA):
    for j in range(colsB):
      prod[i][j] = 0
      for k in range(colsA):
        prod[i][j] += A[i][k]*B[k][j]

  return prod


In [60]:
# Vérification
assert(produit([[1, 2], [3, 4]], [[1, 2], [3, 4]]) == [[7, 10], [15, 22]])

In [61]:
produit(A, A)

[[0.26139703587799, 0.2695652522239007],
 [0.2573332456609382, 0.27259691195164115]]

In [20]:
# Sortons le chronomètre
%timeit produit(A, A)

2.24 µs ± 82.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [14]:
# Et maintenant, avec numpy !
import numpy as np

In [22]:
A2 = np.array(A)
print(A2)

[[0.51844064 0.41890946]
 [0.09178817 0.64490049]]


In [25]:
%timeit np.dot(A, A)

3.71 µs ± 75.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [27]:
for s in [30, 300, 3000]:
  M = np.random.rand(s,s)

  %timeit np.dot(M, M)

  if s == 3000:
    break
  %timeit produit(M, M)



5.7 µs ± 118 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
22.6 ms ± 4.06 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.41 ms ± 838 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
21.5 s ± 476 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.88 s ± 359 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


EXERCICE - Remplissez le tableau suivant avec les durées d'exécution constatées :

(Conseil : pour la taille 3000, essayez UNIQUEMENT avec numpy)

Taille | Python | numpy
-------|--------|-------
30     | 22.6 ms    | 5.7 µs
300    | 21.5 s    | 2.41 ms
3000   | XXX    | 1.88 s

###  Pandas

In [62]:
import pandas as pd

In [63]:
df_exemple = pd.DataFrame({"ESN": ["E420912", "E420913", "E420914"], "panne": [False, False, True]})
df_exemple

,ESN,panne
0,E420912,False
1,E420913,False
2,E420914,True


#### Lecture et prétraitement de données


In [7]:
# !git clone https://github.com/MadaneA/MACS3-Statistiques-Descriptives-TDs.git
!git clone https://github.com/MadaneA/TDs-statistiques-descriptives.git

Cloning into 'TDs-statistiques-descriptives'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 22 (delta 0), reused 22 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (22/22), 21.60 MiB | 7.64 MiB/s, done.


In [ ]:
import os
os.chdir('TDs-statistiques-descriptives')

In [ ]:
os.getcwd()

In [67]:
# Chargement d'un fichier CSV ou Excel
df = pd.read_csv("./data/Vol010.csv")

# Affichage des 5 premières lignes
df.head()

,t,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
0,datenum,deg_C,_,_,%,%,%,%_RPM,psi,_,...,_,_,DEG,_,_,_,_,DEG,lb/h,mach
1,15/09/2011 14:25:58.125,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,15/09/2011 14:25:58.375,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,15/09/2011 14:25:58.625,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,15/09/2011 14:25:58.875,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,7679.98,0.15


In [72]:
"""
EXERCICE - Dimensions d'un DataFrame
Affichez le nombre de colonnes et de ligne du DataFrame (indice : beaucoup de méthodes sont communes entre numpy et pandas)
"""

print('Nombre de colonnes :', df.columns.size)
print('Nombre de lignes :', df.index.size)

Nombre de colonnes : 26
Nombre de lignes : 22945


Les colonnes des DF sont typées, à la manière d'une base de données relationnelle, contrairement aux variables python classiques. Les types des colonnes sont accessibles via df.dtypes. Les principaux types sont les numériques (int32, int64, float etc.

In [73]:
df.loc[10:15, ['EGT_SEL', 'FLIGHT_MOD']]

,EGT_SEL,FLIGHT_MOD
10,0.00234327,NaN
11,1.08032e-007,NaN
12,3.12514e-012,NaN
13,0,NaN
14,72.0006,NaN
15,215.996,NaN


In [74]:
"""
EXERCICE - Extraction et suppression des unités
On remarque que la 1ère ligne ne contient pas de données mais les unités de chaque colonne.
Pour la suite des traitements, il faut supprimer cette ligne. On souhaite toutefois garder
l'information des unités de chaque colonne.
1. Récupérez les unités et stockez les dans une structure adaptée.
2. Supprimez cette ligne du DataFrame en utilisant la méthode "drop"
"""

units = df.iloc[0]
df = df.drop(index=0)

In [75]:
print(units)

df.head()

t             datenum
EGT_SEL         deg_C
FLIGHT_MOD          _
FMV_SEL             _
HPTC_SEL            %
LPTC_SEL            %
N1_SEL              %
N2_ACTSEL       %_RPM
OIL_P             psi
OIL_TEMP            _
PS3_SEL           psi
PT2_SEL            mb
P0_SEL            psi
TAT             deg_C
TBV_SEL             %
TRA_SEL           DEG
T25_SEL             _
T3_SEL              _
VBV_SEL           DEG
VIB_CN1             _
VIB_CN2             _
VIB_TN1             _
VIB_TN2             _
VSV_SEL           DEG
WFM_SEL          lb/h
XM               mach
Name: 0, dtype: object


,t,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
1,15/09/2011 14:25:58.125,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,15/09/2011 14:25:58.375,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,15/09/2011 14:25:58.625,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,15/09/2011 14:25:58.875,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,7679.98,0.15
5,15/09/2011 14:25:59.125,NaN,NaN,NaN,NaN,NaN,0,0,647.998,NaN,...,NaN,NaN,0,0,0,0,0,0,4607.86,0.15


On remarque que toutes les colonnes ont été reconnues comme de type object, c'est-à-dire des chaînes de caractères, alors que ce sont des valeurs numériques. Cela est dû à la première ligne contenant les unités. Il faut donc convertir les colonnes en numérique. La colonne 't', quant à elle, doit être convertie en type datetime.

In [76]:
df['t'] = pd.to_datetime(df['t'])
df[df.columns[1:]] = df[df.columns[1:]].apply(pd.to_numeric)

/tmp/ipykernel_32150/2306591669.py:1: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S.%f format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['t'] = pd.to_datetime(df['t'])


In [77]:
df.dtypes

t             datetime64[ns]
EGT_SEL              float64
FLIGHT_MOD           float64
FMV_SEL              float64
HPTC_SEL             float64
LPTC_SEL             float64
N1_SEL               float64
N2_ACTSEL            float64
OIL_P                float64
OIL_TEMP             float64
PS3_SEL              float64
PT2_SEL              float64
P0_SEL               float64
TAT                  float64
TBV_SEL              float64
TRA_SEL              float64
T25_SEL              float64
T3_SEL               float64
VBV_SEL              float64
VIB_CN1              float64
VIB_CN2              float64
VIB_TN1              float64
VIB_TN2              float64
VSV_SEL              float64
WFM_SEL              float64
XM                   float64
dtype: object

In [91]:
"""
Exercice - Index temporel
Comme nos données sont une série temporelle multivariée, on souhaite utiliser un index temporel.
1. Créez une copie de df, appelée df2, à l'aide de la méthode du même nom.
2. Affectez la colonne du temps ('t') en tant qu'indice du DataFrame.
3. Supprimez la colonne 't' du DF résultant.
"""

df2 = df.copy()
# df2.index = df2["t"]
# df = df2.drop(columns="t")
df2 = df2.set_index("t")

df2


,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,PS3_SEL,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
t,,,,,,,,,,,,,,,,,,,,,
2011-09-15 14:25:58.125,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.375,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.625,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.875,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,7679.98,0.15
2011-09-15 14:25:59.125,NaN,NaN,NaN,NaN,NaN,0.0,0.0,647.998,NaN,575.982,...,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,4607.86,0.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2011-09-15 16:01:32.875,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 16:01:33.125,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 16:01:33.375,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


On constate que pandas a automatiquement reconnu un DatetimeIndex, adapté pour des manipulations de séries temporelles (moyennes glissantes, etc) !

### Valeurs manquantes

**NaN = Not a Number**

Les valeurs NaN doivent être éliminées ou imputées (i.e. remplacées par une certaine valeur) avant la suite des traitements. Ce choix dépend du cas d'usage. Dans un premier temps, nous allons apprendre à :

* trouver les données manquantes (méthode `isna`)
* éliminer les données manquantes d'un DataFrame (méthode `dropna`)
* les remplacer par une constante (méthode (`fillna`)

In [92]:
"""
EXERCICE - La méthode isna
1. Testez la méthode isna sur le DataFrame df2, puis sur une colonne ou une ligne. Que renvoie-t-elle ?
2. En appliquant les méthodes any(axis=...), mean() et max()/idxmax() sur les résultats de isna(), répondez aux questions suivantes :
    2.1 Quelles colonnes contiennent des valeurs manquantes, lesquelles n'en contiennent pas ?
    2.2 Quel est le pourcentage de valeurs manquantes dans le DF (a) par colonne (b) globalement ? Quelle variable contient le plus de NaN ?
    2.3 Quel est le pourcentage d'indices du DF pour lesquels toutes les variables sont présentes ?
"""

print()

#### Q1 Ans:

The method returns a dataframe/array of boolean values, as can be seen.

In [93]:
df2.isna()

,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,PS3_SEL,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
t,,,,,,,,,,,,,,,,,,,,,
2011-09-15 14:25:58.125,True,True,True,True,True,True,False,True,True,True,...,True,True,True,True,True,True,True,True,True,True
2011-09-15 14:25:58.375,True,True,True,True,True,True,False,True,True,True,...,True,True,True,True,True,True,True,True,True,True
2011-09-15 14:25:58.625,True,True,True,True,True,True,False,True,True,True,...,True,True,True,True,True,True,True,True,True,True
2011-09-15 14:25:58.875,True,True,True,True,True,True,False,True,True,True,...,True,True,True,True,True,True,True,True,False,False
2011-09-15 14:25:59.125,True,True,True,True,True,False,False,False,True,False,...,True,True,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2011-09-15 16:01:32.875,True,True,True,True,True,True,True,True,True,True,...,True,True,True,True,True,True,True,True,True,True
2011-09-15 16:01:33.125,True,True,True,True,True,True,True,True,True,True,...,True,True,True,True,True,True,True,True,True,True
2011-09-15 16:01:33.375,True,True,True,True,True,True,True,True,True,True,...,True,True,True,True,True,True,True,True,True,True


In [94]:
df2["EGT_SEL"].isna()

t
2011-09-15 14:25:58.125    True
2011-09-15 14:25:58.375    True
2011-09-15 14:25:58.625    True
2011-09-15 14:25:58.875    True
2011-09-15 14:25:59.125    True
                           ... 
2011-09-15 16:01:32.875    True
2011-09-15 16:01:33.125    True
2011-09-15 16:01:33.375    True
2011-09-15 16:01:33.625    True
2011-09-15 16:01:33.875    True
Name: EGT_SEL, Length: 22944, dtype: bool

#### Q2 Ans

#### Question 2.1:

by using the methods, all columns seem to have missing values.

In [95]:
df2.isna().any(axis=0)

EGT_SEL       True
FLIGHT_MOD    True
FMV_SEL       True
HPTC_SEL      True
LPTC_SEL      True
N1_SEL        True
N2_ACTSEL     True
OIL_P         True
OIL_TEMP      True
PS3_SEL       True
PT2_SEL       True
P0_SEL        True
TAT           True
TBV_SEL       True
TRA_SEL       True
T25_SEL       True
T3_SEL        True
VBV_SEL       True
VIB_CN1       True
VIB_CN2       True
VIB_TN1       True
VIB_TN2       True
VSV_SEL       True
WFM_SEL       True
XM            True
dtype: bool

#### Question 2.2

##### (a) the percentages by column are as follows:

In [96]:
df2.isna().mean(axis=0)*100

EGT_SEL       0.348675
FLIGHT_MOD    0.444561
FMV_SEL       0.444561
HPTC_SEL      0.422768
LPTC_SEL      0.457636
N1_SEL        0.418410
N2_ACTSEL     0.252789
OIL_P         0.379184
OIL_TEMP      0.383543
PS3_SEL       0.300732
PT2_SEL       0.370467
P0_SEL        0.370467
TAT           0.300732
TBV_SEL       0.296374
TRA_SEL       0.300732
T25_SEL       0.470711
T3_SEL        0.470711
VBV_SEL       0.292015
VIB_CN1       0.292015
VIB_CN2       0.283298
VIB_TN1       0.283298
VIB_TN2       0.283298
VSV_SEL       0.274582
WFM_SEL       0.270223
XM            0.261506
dtype: float64

##### (b) the overall percentage of mising values is:

In [97]:
df2.isna().mean().mean()*100

np.float64(0.346931659693166)

and the variable with most NaNs is the following (it does not consider ties):

In [85]:
df2.isna().mean().idxmax()

'T25_SEL'

the variables with most NaNs are the following (now considers ties):

In [86]:
df2.columns[df2.isna().mean() == df2.isna().mean().max()]

Index(['T25_SEL', 'T3_SEL'], dtype='object')

#### Question 2.3

The percentage of indices in the DF for which all variables are present is:

In [87]:
(1-df2.isna().any(axis=1)).mean()*100

np.float64(99.48134588563458)

In [54]:
"""
EXERCICE - La méthode dropna
La méthode dropna permet d'éliminer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. À quoi correspondent les arguments "axis" et "how" ?
2. Éliminez toutes les lignes contenant uniquement des valeurs manquantes.
3. Éliminez toutes les lignes contenant au moins une valeur manquante. Combien y a-t-il de lignes de différence ?
4. Éliminez toutes les colonnes contenant au moins une valeur manquante.
"""

print()

#### Q1 Ans:

The 'axis' method determines the 'direction' in which rows/columns with NaNs are removed - 0 for rows and 1 for cols. The 'how' method describes the rule to drop the row/col - 'any' removes the index if any NaN appears, 'all' removes if all values are NaNs.

#### Q2 Ans:

We do that as follows:

In [101]:
df2_dropna_rows = df2.dropna(axis=0, how="all")
df2_dropna_rows

,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,PS3_SEL,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
t,,,,,,,,,,,,,,,,,,,,,
2011-09-15 14:25:58.125,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.375,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.625,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2011-09-15 14:25:58.875,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,7679.980000,0.15
2011-09-15 14:25:59.125,NaN,NaN,NaN,NaN,NaN,0.0,0.0,647.998,NaN,575.982,...,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,4607.860000,0.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2011-09-15 16:01:23.625,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,0.008331,0.15
2011-09-15 16:01:23.875,NaN,NaN,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,...,NaN,NaN,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.15
2011-09-15 16:01:24.125,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,0.0,0.0,0.0,0.0,0.000000,0.15


#### Q3 Ans:

In [ ]:
df2_rows = df2.index.size
df2_dropna = df2.dropna(axis=0, how="any")
df2_dropna_rows = df2_dropna.index.size

print("before dropping, there were {} rows; after dropping there are {} rows."
      "the difference is of {} rows".format(df2_rows, df2_dropna_rows, df2_rows-df2_dropna_rows))

before dropping, there were 22944 rows; after dropping there are 22825 rows. the difference is of 119 rows


#### Q4 Ans

In [103]:
df2.dropna(axis=0, how="any")

,EGT_SEL,FLIGHT_MOD,FMV_SEL,HPTC_SEL,LPTC_SEL,N1_SEL,N2_ACTSEL,OIL_P,OIL_TEMP,PS3_SEL,...,T25_SEL,T3_SEL,VBV_SEL,VIB_CN1,VIB_CN2,VIB_TN1,VIB_TN2,VSV_SEL,WFM_SEL,XM
t,,,,,,,,,,,,,,,,,,,,,
2011-09-15 14:26:02.875,335.9840,1.0,0.000,0.0,0.0,0.0,0.0,0.0,177.3740,959.98000,...,-128.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,7679.84,0.15
2011-09-15 14:26:03.125,239.9860,1.0,0.000,0.0,0.0,0.0,0.0,0.0,133.6240,575.98200,...,-128.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,4607.86,0.15
2011-09-15 14:26:03.375,167.9910,1.0,0.000,0.0,0.0,0.0,0.0,0.0,89.8743,191.98400,...,-128.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1535.87,0.15
2011-09-15 14:26:03.625,119.9940,1.0,0.000,0.0,0.0,0.0,0.0,0.0,68.0000,0.00000,...,-128.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.00,0.15
2011-09-15 14:26:03.875,72.0011,1.0,0.000,0.0,0.0,0.0,0.0,0.0,68.0000,0.00000,...,-128.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.00,0.15
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2011-09-15 16:01:19.875,230.8250,1.0,-12.375,0.0,0.0,0.0,0.0,0.0,69.0000,7.50011,...,34.5,178.0,0.0,0.0,0.0,0.0,0.0,0.0,0.00,0.15
2011-09-15 16:01:20.125,137.4770,1.0,-12.375,0.0,0.0,0.0,0.0,0.0,69.0000,4.50085,...,34.5,178.0,0.0,0.0,0.0,0.0,0.0,0.0,0.00,0.15
2011-09-15 16:01:20.375,113.4530,1.0,-12.375,0.0,0.0,0.0,0.0,0.0,69.0000,1.50063,...,34.5,178.0,0.0,0.0,0.0,0.0,0.0,0.0,0.00,0.15


In [41]:
"""
EXERCICE - La méthode fillna
La méthode dropna permet d'imputer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. Quelles sont les différentes stratégies de remplissage des valeurs manquantes ?
2. Imputez les valeurs manquantes de la colonne age du DF donné en exemple par :
    - 0
    - la dernière valeur précédente/suivante valide
    - la moyenne
    - la valeur la plus courante (mode)
3. Quel est le meilleur choix dans ce cas ? Et pour le cas d'une variable temporelle, par exemple la température 'EGT_SEL' de notre jeu de données ?
"""
exemple = pd.DataFrame({'nom': ['Alice', 'Bob', 'Charlie', 'David'], 'age': [24, np.nan, 99, 24]})

#### Q1 Ans:

The pandas.DataFrame.fillna method has many strategies and parameters describing different manners to fill missing values: the field 'value' allows one to parse a fixed value, or structured data to fill gaps; the field 'method' has options to fill a missing value by making use of the last/next valid value; also, the field named 'axis' receives the direction to wich consider when filling missing values.

#### Q2 Ans:

We do as follows for th different requirements:

In [45]:
exemple_0 = exemple.fillna(value=0)
exemple_0

,nom,age
0,Alice,24.0
1,Bob,0.0
2,Charlie,99.0
3,David,24.0


In [48]:
exemple_1_ffill = exemple.ffill() # used ffill method because fillna(method="ffill") is depracated
exemple_1_ffill

,nom,age
0,Alice,24.0
1,Bob,24.0
2,Charlie,99.0
3,David,24.0


In [49]:
exemple_1_bfill = exemple.bfill() # used bfill method because fillna(method="bfill") is depracated
exemple_1_bfill

,nom,age
0,Alice,24.0
1,Bob,99.0
2,Charlie,99.0
3,David,24.0


In [50]:
example_2 = exemple.fillna(value=exemple["age"].mean())
example_2

,nom,age
0,Alice,24.0
1,Bob,49.0
2,Charlie,99.0
3,David,24.0


In [51]:
example_3 = exemple.fillna(value=exemple.mode().iloc[0])
example_3

,nom,age
0,Alice,24.0
1,Bob,24.0
2,Charlie,99.0
3,David,24.0


#### Q3 Ans:

In fact, a choice of strategy for filling missing data depends strongly on the data and use. In this case, however, a good choice seems to be the average, since there is no relation between individuals that justify the use of a positional strategy, as 'bfill' and 'ffill'. Filling the missing field with the average value, on the other side, avoids changing the statistical distribution of the data in question.

On the other side, a temporal variable might (probably) has a temporal dependency, in this case, filling a missing field that varies over time, such as 'EGT_SEL' in our dataset, might require an strategy tha takes into consideration the time/sequential structure. Filling a missing temperature with the average observed may not be representative; instead, filling sich value with the previous/following or even the average of both may be a better option, especially if we assume the temperature transition to be smooth.

## Git

Créer un compte GitHub : https://github.com/

Créer un nouveau repository :
1. Sélectionner "Nouveau dépôt" dans le menu déroulant avec le signe +.
2. Saisisser un nom pour votre dépôt (par exemple, "TDs statistiques descriptives")
3. Cliquer sur "Créer un dépôt". Ne vous souciez pas des autres options.



Télécharger Git Bash : https://git-scm.com/downloads

Sur Git Bash, positionnez-vous dans le répertoire où votre projet figure.

Ensuite

> git init

> git remote add origin https://github.com/********

> git add .

> git commit -m "Ajout du TD1"

> git push origin master

C'est tout ! Vous avez créé votre repo GitHub !